### PR TITLE
Enable SCRAM-SHA-256 and SCRAM-SHA-512 for sasl

### LINKED ISSUE #1882: Support for sasl mechanism: SCRAM-SHA-256
Is there any chance for having SCRAM-SHA-256 mechanism supported in kafka-python?

### PR DESCRIPTION
**Pull request by swenzel**
closes #1882

---
This change is [Reviewable](https://reviewable.io/reviews/dpkp/kafka-python/1918)

### CODE REVIEW

**Comment by swenzel** [comment]
Documentation is still missing, I'll add something on Friday.

**Comment by swenzel** [comment]
For some reason my changes to the `kafka.properties` files seem to be ignored. I have no clue what is going on there. I'm out of ideas.

**Comment by swenzel** [comment]
Wow, that one took me a while...
Okay the issue was the travis cache which was caching all resource directories. Two solutions to that:
1. Keep my current solution which only caches `servers/dist` meaning that the archives have to be unpacked all the time but that doesn't take long.
2. Run `rm servers/*/resources` during https://docs.travis-ci.com/user/caching/#before_cache-phase to not add the resources to the cache.

**Comment by swenzel** [comment]
@dpkp @jeffwidman ready to be reviewed now.
That one test that is still failing seems to only fail some of the times. My best guess is that the broker itself hasn't completely fetched the data yet and that a small `sleep` might fix this.
But I think I have already fixed more than I should :sweat_smile: 
Looking forward to reading your feedback :)

**Comment by ofek** [comment]
@dpkp @jeffwidman We are extremely interested in this as well. Could you please take a look?

**Comment by israelglar** [comment]
Would be great to have this feature. Is there a timeframe for when we can merge it? 

Thanks!

File: .travis.yml [hunk 1]
```diff
@@ -25,7 +25,7 @@ addons:
 cache:
   directories:
     - $HOME/.cache/pip
-    - servers/
+    - servers/dist
```
Comment (dpkp) [inline .travis.yml]: good catch!

File: kafka/client_async.py [hunk 2]
```diff
@@ -767,10 +767,7 @@ def least_loaded_node(self):
                 inflight = curr_inflight
                 found = node_id
 
-        if found is not None:
-            return found
-
-        return None
+        return found
```
Comment (dpkp) [inline kafka/client_async.py]: drive-by!

File: test/__init__.py [hunk 3]
```diff
@@ -2,14 +2,7 @@
...
 import logging
-try:  # Python 2.7+
-    from logging import NullHandler
-except ImportError:
-    class NullHandler(logging.Handler):
-        def emit(self, record):
-            pass
-
-logging.getLogger(__name__).addHandler(NullHandler())
+logging.basicConfig(level=logging.INFO)
```
Comment (dpkp) [inline test/__init__.py]: What's your thinking on the logging changes here?

File: kafka/conn.py [hunk 4]
```diff
@@ -98,6 +107,69 @@ class ConnectionStates(object):
     AUTHENTICATING = '<authenticating>'
 
 
+class ScramClient:
```
Comment (dpkp) [inline kafka/conn.py]: Thoughts on moving this to its own file ?

File: kafka/conn.py [hunk 5]
```diff
@@ -652,6 +730,53 @@ def _try_authenticate_plain(self, future):
...
+
+        err = None
+        close = False
+        with self._lock:
+            if not self._can_send_recv():
+                err = Errors.NodeNotReadyError(str(self))
+                close = False
+            else:
+                try:
+                    client_first = scram_client.first_message().encode()
```
Comment (dpkp) [inline kafka/conn.py]: should we specify an encoding explicitly? is 'utf-8' always the default? does it matter?

File: kafka/conn.py [hunk 6]
```diff
@@ -652,6 +730,53 @@ def _try_authenticate_plain(self, future):
...
+                    size = Int32.encode(len(client_first))
+                    self._send_bytes_blocking(size + client_first)
+
+                    (data_len,) = struct.unpack('>i', self._recv_bytes_blocking(4))
+                    server_first = self._recv_bytes_blocking(data_len).decode()
```
Comment (dpkp) [inline kafka/conn.py]: similarly re: codec. would prefer to specific 'utf-8' explicitly

File: kafka/conn.py [hunk 7]
```diff
@@ -652,6 +730,53 @@ def _try_authenticate_plain(self, future):
...
+                    scram_client.process_server_first_message(server_first)
+
+                    client_final = scram_client.final_message().encode()
```
Comment (dpkp) [inline kafka/conn.py]: and here - utf-8

File: kafka/conn.py [hunk 8]
```diff
@@ -652,6 +730,53 @@ def _try_authenticate_plain(self, future):
...
+                    size = Int32.encode(len(client_final))
+                    self._send_bytes_blocking(size + client_final)
+
+                    (data_len,) = struct.unpack('>i', self._recv_bytes_blocking(4))
+                    server_final = self._recv_bytes_blocking(data_len).decode()
```
Comment (dpkp) [inline kafka/conn.py]: and finally also here

File: test/fixtures.py [hunk 9]
```diff
@@ -271,13 +285,18 @@ def __init__(self, host, port, broker_id, zookeeper, zk_chroot,
         self.broker_id = broker_id
         self.auto_create_topic = auto_create_topic
         self.transport = transport.upper()
-        self.sasl_mechanism = sasl_mechanism.upper()
+        if sasl_mechanism is not None:
+            self.sasl_mechanism = sasl_mechanism.upper()
+        else:
+            self.sasl_mechanism = None
```
Comment (dpkp) [inline test/fixtures.py]: the else clause seems redundant?

File: test/fixtures.py [hunk 10]
```diff
@@ -446,80 +550,122 @@ def _failure(error):
...
-                error_code = topic_result[1]
-                if error_code != 0:
-                    raise errors.for_code(error_code)
+            try:
+                self._create_topic_via_admin_api(topic_name, num_partitions, replication_factor, timeout_ms)
+            except InvalidReplicationFactorError:
+                # wait and try again
+                # on travis the brokers sometimes take a while to find themselves
+                time.sleep(0.5)
+                self._create_topic_via_admin_api(topic_name, num_partitions, replication_factor, timeout_ms)
```
Comment (dpkp) [inline test/fixtures.py]: might prefer an outer loop to enable as many retries as needed until timeout

**Comment by ofek** [comment]
@dpkp Hello! Will this be released soon?

**Comment by ofek** [comment]
@dpkp @jeffwidman Any update?

**Comment by ofek** [comment]
Alternatively, can I please do something to help?

**Comment by jeffwidman** [comment]
Hey @ofek I don't have the rights to cut a new release, but I just pinged @dpkp and he said on IM that he's currently traveling w/o his laptop for the next week so can't look at it until he returns. Sorry about that.

**Comment by ofek** [comment]
No worries, thanks!

**Comment by ofek** [comment]
Hello again! Any update?

**Comment by jfpatenaude** [comment]
I'm also waiting on this feature, @dpkp can you give us an update on the expected timeline?

**Comment by terence-bigtt** [comment]
looking forward for that too !

**Comment by ofek** [comment]
@terence-bigtt This is released.

**Comment by Neustradamus** [comment]
@swenzel, @dpkp: Thanks a lot for all about SCRAM.

Can you look, a guy has found a bug?
- https://github.com/dpkp/kafka-python/issues/2238

Linked to:
- https://github.com/scram-xmpp/info/issues/1
